File: src/main/java/org/cubexmc/metro/service/LineCommandService.java
package org.cubexmc.metro.service;

import java.util.List;
import java.util.UUID;
import java.util.Comparator;
import java.util.regex.Pattern;

import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.model.EntityModelController;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.model.LineStatus;
import org.cubexmc.metro.model.Portal;
import org.cubexmc.metro.model.PriceRule;
import org.cubexmc.metro.model.Stop;
import org.cubexmc.metro.event.LineStatusChangeEvent;
import org.bukkit.Bukkit;

/**
 * Business operations used by line commands.
 */
public class LineCommandService {

    private static final int MAX_ID_LENGTH = 64;
    private static final Pattern ID_PATTERN = Pattern.compile("[A-Za-z0-9_-]+");
    private static final Pattern LEGACY_COLOR_PATTERN = Pattern.compile("(?i)&[0-9A-FK-OR]");
    private static final Pattern HEX_COLOR_PATTERN = Pattern.compile("(?i)&#[0-9A-F]{6}");

    private final LineManager lineManager;

    public LineCommandService(LineManager lineManager) {
        this.lineManager = lineManager;
    }

    public enum WriteStatus {
        SUCCESS,
        INVALID_ID,
        INVALID_COLOR,
        INVALID_VALUE,
        EXISTS,
        NOT_FOUND,
        FAILED,
        STOP_NO_WORLD,
        WORLD_MISMATCH,
        CIRCULAR_INVALID_INDEX
    }

    public record AddStopResult(WriteStatus status, String lineWorld, String stopWorld) {
        static AddStopResult of(WriteStatus status) {
            return new AddStopResult(status, null, null);
        }

        static AddStopResult worldMismatch(String lineWorld, String stopWorld) {
            return new AddStopResult(WriteStatus.WORLD_MISMATCH, lineWorld, stopWorld);
        }
    }

    public record ClearRouteResult(WriteStatus status, int previousPointCount) {
    }

    public WriteStatus createLine(String id, String name, UUID ownerId) {
        if (!isValidId(id)) {
            return WriteStatus.INVALID_ID;
        }
        return lineManager.createLine(id, name, ownerId) ? WriteStatus.SUCCESS : WriteStatus.EXISTS;
    }

    public WriteStatus deleteLine(String id) {
        return lineManager.deleteLine(id) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public List<Line> listLines() {
        return lineManager.getAllLines().stream()
                .sorted(Comparator.comparing(Line::getId))
                .toList();
    }

    public WriteStatus renameLine(String id, String name) {
        return lineManager.setLineName(id, name) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus setColor(String id, String color) {
        if (!isValidColor(color)) {
            return WriteStatus.INVALID_COLOR;
        }
        return lineManager.setLineColor(id, color) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus setTerminusName(String id, String terminusName) {
        return lineManager.setLineTerminusName(id, terminusName) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus setMaxSpeed(String id, double speed) {
        if (!Double.isFinite(speed) || speed <= 0.0) {
            return WriteStatus.INVALID_VALUE;
        }
        return lineManager.setLineMaxSpeed(id, speed) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus setEntityType(String id, String entityTypeRaw) {
        String normalized = EntityModelController.normalizeLineEntityType(entityTypeRaw);
        if (normalized == null) {
            return WriteStatus.INVALID_VALUE;
        }
        return lineManager.setLineEntityType(id, normalized) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public AddStopResult addStopToLine(Line line, Stop stop, Integer index) {
        String stopWorld = stop.getWorldName();
        String lineWorld = line.getWorldName();
        if (stopWorld == null || stopWorld.isBlank()) {
            return AddStopResult.of(WriteStatus.STOP_NO_WORLD);
        }
        if (lineWorld != null && !lineWorld.equals(stopWorld)) {
            return AddStopResult.worldMismatch(lineWorld, stopWorld);
        }

        int targetIndex = index == null ? -1 : index;
        List<String> orderedStopIds = line.getOrderedStopIds();
        if (line.isCircular() && (targetIndex < 0 || targetIndex >= orderedStopIds.size())) {
            return AddStopResult.of(WriteStatus.CIRCULAR_INVALID_INDEX);
        }

        if (!lineManager.addStopToLine(line.getId(), stop.getId(), targetIndex)) {
            return AddStopResult.of(WriteStatus.FAILED);
        }
        if (lineWorld == null) {
            lineManager.setLineWorldName(line.getId(), stopWorld);
        }
        return AddStopResult.of(WriteStatus.SUCCESS);
    }

    public WriteStatus removeStopFromLine(Line line, String stopId) {
        if (!lineManager.delStopFromLine(line.getId(), stopId)) {
            return WriteStatus.FAILED;
        }
        if (line.getOrderedStopIds().isEmpty()) {
            lineManager.setLineWorldName(line.getId(), null);
        }
        return WriteStatus.SUCCESS;
    }

    public WriteStatus addPortalToLine(Line line, Portal portal) {
        if (line == null || portal == null) {
            return WriteStatus.FAILED;
        }
        if (line.containsPortal(portal.getId())) {
            return WriteStatus.EXISTS;
        }
        return lineManager.addPortalToLine(line.getId(), portal.getId()) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus removePortalFromLine(Line line, String portalId) {
        if (line == null || portalId == null || portalId.isBlank()) {
            return WriteStatus.FAILED;
        }
        if (!line.containsPortal(portalId)) {
            return WriteStatus.NOT_FOUND;
        }
        return lineManager.delPortalFromLine(line.getId(), portalId) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus setRailProtected(String id, boolean enabled) {
        return lineManager.setLineRailProtected(id, enabled) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus grantAdmin(Line line, UUID adminId) {
        if (adminId == null) {
            return WriteStatus.INVALID_VALUE;
        }
        if (line.getAdmins().contains(adminId)) {
            return WriteStatus.EXISTS;
        }
        return lineManager.addLineAdmin(line.getId(), adminId) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus revokeAdmin(Line line, UUID adminId) {
        if (adminId == null) {
            return WriteStatus.INVALID_VALUE;
        }
        return lineManager.removeLineAdmin(line.getId(), adminId) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus transferOwner(Line line, UUID ownerId) {
        if (ownerId == null) {
            return WriteStatus.INVALID_VALUE;
        }
        return lineManager.setLineOwner(line.getId(), ownerId) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public ClearRouteResult clearRoutePoints(Line line) {
        int previousCount = line.getRoutePoints().size();
        WriteStatus status = lineManager.clearLineRoutePoints(line.getId()) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
        return new ClearRouteResult(status, previousCount);
    }

    public WriteStatus cloneReverseLine(String sourceId, String newId, String stopIdSuffix, UUID ownerId) {
        if (!isValidId(newId)) {
            return WriteStatus.INVALID_ID;
        }
        return lineManager.cloneReverseLine(sourceId, newId, stopIdSuffix, ownerId)
                ? WriteStatus.SUCCESS
                : WriteStatus.FAILED;
    }

    public WriteStatus setTicketPrice(String id, double price) {
        if (!Double.isFinite(price) || price < 0.0) {
            return WriteStatus.INVALID_VALUE;
        }
        return lineManager.setLineTicketPrice(id, price) ? WriteStatus.SUCCESS : WriteStatus.FAILED;
    }

    public WriteStatus setPriceRule(String id, String mode, double basePrice, Double perUnit, Double maxPrice) {
        if (basePrice < 0.0 || (perUnit != null && perUnit < 0.0) || (maxPrice != null && maxPrice < 0.0)) {
            return WriteStatus.INVALID_VALUE;
        }

        PriceRule.PricingMode pricingMode;
        try {
            pricingMode = PriceRule.PricingMode.valueOf(mode.toUpperCase());
        } catch (IllegalArgumentException e) {
            return WriteStatus.INVALID_VALUE;
        }

        Line line = lineManager.getLine(id);
        if (line == null) {
            return WriteStatus.NOT_FOUND;
        }

        PriceRule rule = new PriceRule(pricingMode, basePrice);
        if (pricingMode == PriceRule.PricingMode.DISTANCE && perUnit != null) {
            rule.setPerBlockRate(perUnit);
        }
        if (pricingMode == PriceRule.PricingMode.INTERVAL && perUnit != null) {
            rule.setPerIntervalRate(perUnit);
        }
        if (maxPrice != null && maxPrice > 0.0) {
            rule.setMaxPrice(maxPrice);
        }

        line.setPriceRule(rule);
        lineManager.saveConfig();
        return WriteStatus.SUCCESS;
    }

    public boolean resetPriceRule(String id) {
        Line line = lineManager.getLine(id);
        if (line == null) return false;
        line.setPriceRule(null);
        lineManager.saveConfig();
        return true;
    }

    public WriteStatus setLineStatus(String id, String status) {
        LineStatus lineStatus = LineStatus.fromConfig(status);
        if (lineStatus == LineStatus.NORMAL && !status.equalsIgnoreCase("normal")) {
            return WriteStatus.INVALID_VALUE;
        }

        Line line = lineManager.getLine(id);
        if (line == null) {
            return WriteStatus.NOT_FOUND;
        }

        LineStatus oldStatus = line.getLineStatus();
        if (oldStatus == lineStatus) {
            return WriteStatus.SUCCESS;
        }

        line.setLineStatus(lineStatus);
        Bukkit.getPluginManager().callEvent(new LineStatusChangeEvent(line, oldStatus, lineStatus));
        lineManager.saveConfig();
        return WriteStatus.SUCCESS;
    }

    public boolean addAlternativeRoute(String id, String altLineId) {
        Line line = lineManager.getLine(id);
        if (line == null) return false;
        boolean result = line.addAlternativeRoute(altLineId);
        if (result) {
            lineManager.saveConfig();
        }
        return result;
    }

    public boolean removeAlternativeRoute(String id, String altLineId) {
        Line line = lineManager.getLine(id);
        if (line == null) return false;
        boolean result = line.removeAlternativeRoute(altLineId);
        if (result) {
            lineManager.saveConfig();
        }
        return result;
    }

    public boolean setSuspensionMessage(String id, String message) {
        Line line = lineManager.getLine(id);
        if (line == null) return false;
        line.setSuspensionMessage(message);
        lineManager.saveConfig();
        return true;
    }

    public boolean isValidId(String id) {
        return id != null
                && !id.isBlank()
                && id.length() <= MAX_ID_LENGTH
                && ID_PATTERN.matcher(id).matches();
    }

    public boolean isValidColor(String color) {
        return color != null
                && (LEGACY_COLOR_PATTERN.matcher(color).matches()
                || HEX_COLOR_PATTERN.matcher(color).matches());
    }
}

File: src/main/java/org/cubexmc/metro/service/LineCommandService.kt
package org.cubexmc.metro.service

import java.util.Collections
import java.util.Locale
import java.util.UUID
import java.util.regex.Pattern
import org.bukkit.Bukkit
import org.cubexmc.metro.event.LineStatusChangeEvent
import org.cubexmc.metro.manager.LineManager
import org.cubexmc.metro.model.EntityModelController
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.LineStatus
import org.cubexmc.metro.model.Portal
import org.cubexmc.metro.model.PriceRule
import org.cubexmc.metro.model.Stop

/**
 * Business operations used by line commands.
 */
class LineCommandService(lineManager: LineManager?) {
    private val lineManagerRef: LineManager? = lineManager
    private val lineManager: LineManager
        get() = lineManagerRef ?: throw NullPointerException("lineManager")

    enum class WriteStatus {
        SUCCESS,
        INVALID_ID,
        INVALID_COLOR,
        INVALID_VALUE,
        EXISTS,
        NOT_FOUND,
        FAILED,
        STOP_NO_WORLD,
        WORLD_MISMATCH,
        CIRCULAR_INVALID_INDEX,
    }

    @JvmRecord
    data class AddStopResult(
        val status: WriteStatus,
        val lineWorld: String?,
        val stopWorld: String?,
    )

    @JvmRecord
    data class ClearRouteResult(
        val status: WriteStatus,
        val previousPointCount: Int,
    )

    fun createLine(id: String?, name: String?, ownerId: UUID?): WriteStatus {
        if (!isValidId(id)) {
            return WriteStatus.INVALID_ID
        }
        return if (lineManager.createLine(id, name, ownerId)) WriteStatus.SUCCESS else WriteStatus.EXISTS
    }

    fun deleteLine(id: String): WriteStatus =
        if (lineManager.deleteLine(id)) WriteStatus.SUCCESS else WriteStatus.FAILED

    fun listLines(): List<Line> = Collections.unmodifiableList(
        lineManager.getAllLines().sortedBy { line -> line.id },
    )

    fun renameLine(id: String, name: String?): WriteStatus =
        if (lineManager.setLineName(id, name)) WriteStatus.SUCCESS else WriteStatus.FAILED

    fun setColor(id: String, color: String?): WriteStatus {
        if (!isValidColor(color)) {
            return WriteStatus.INVALID_COLOR
        }
        return if (lineManager.setLineColor(id, color)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun setTerminusName(id: String, terminusName: String?): WriteStatus =
        if (lineManager.setLineTerminusName(id, terminusName)) WriteStatus.SUCCESS else WriteStatus.FAILED

    fun setMaxSpeed(id: String, speed: Double): WriteStatus {
        if (!java.lang.Double.isFinite(speed) || speed <= 0.0) {
            return WriteStatus.INVALID_VALUE
        }
        return if (lineManager.setLineMaxSpeed(id, speed)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun setEntityType(id: String, entityTypeRaw: String?): WriteStatus {
        val normalized = EntityModelController.normalizeLineEntityType(entityTypeRaw)
            ?: return WriteStatus.INVALID_VALUE
        return if (lineManager.setLineEntityType(id, normalized)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun addStopToLine(line: Line, stop: Stop, index: Int?): AddStopResult {
        val stopWorld = stop.worldName
        val lineWorld = line.worldName
        if (stopWorld == null || stopWorld.isBlank()) {
            return AddStopResult(WriteStatus.STOP_NO_WORLD, null, null)
        }
        if (lineWorld != null && lineWorld != stopWorld) {
            return AddStopResult(WriteStatus.WORLD_MISMATCH, lineWorld, stopWorld)
        }

        val targetIndex = index ?: -1
        val orderedStopIds = line.orderedStopIds
        if (line.isCircular && (targetIndex < 0 || targetIndex >= orderedStopIds.size)) {
            return AddStopResult(WriteStatus.CIRCULAR_INVALID_INDEX, null, null)
        }

        if (!lineManager.addStopToLine(line.id, stop.id, targetIndex)) {
            return AddStopResult(WriteStatus.FAILED, null, null)
        }
        if (lineWorld == null) {
            lineManager.setLineWorldName(line.id, stopWorld)
        }
        return AddStopResult(WriteStatus.SUCCESS, null, null)
    }

    fun removeStopFromLine(line: Line, stopId: String): WriteStatus {
        if (!lineManager.delStopFromLine(line.id, stopId)) {
            return WriteStatus.FAILED
        }
        if (line.orderedStopIds.isEmpty()) {
            lineManager.setLineWorldName(line.id, null)
        }
        return WriteStatus.SUCCESS
    }

    fun addPortalToLine(line: Line?, portal: Portal?): WriteStatus {
        if (line == null || portal == null) {
            return WriteStatus.FAILED
        }
        if (line.containsPortal(portal.id)) {
            return WriteStatus.EXISTS
        }
        return if (lineManager.addPortalToLine(line.id, portal.id)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun removePortalFromLine(line: Line?, portalId: String?): WriteStatus {
        if (line == null || portalId == null || portalId.isBlank()) {
            return WriteStatus.FAILED
        }
        if (!line.containsPortal(portalId)) {
            return WriteStatus.NOT_FOUND
        }
        return if (lineManager.delPortalFromLine(line.id, portalId)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun setRailProtected(id: String, enabled: Boolean): WriteStatus =
        if (lineManager.setLineRailProtected(id, enabled)) WriteStatus.SUCCESS else WriteStatus.FAILED

    fun grantAdmin(line: Line, adminId: UUID?): WriteStatus {
        if (adminId == null) {
            return WriteStatus.INVALID_VALUE
        }
        if (line.admins.contains(adminId)) {
            return WriteStatus.EXISTS
        }
        return if (lineManager.addLineAdmin(line.id, adminId)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun revokeAdmin(line: Line, adminId: UUID?): WriteStatus {
        if (adminId == null) {
            return WriteStatus.INVALID_VALUE
        }
        return if (lineManager.removeLineAdmin(line.id, adminId)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun transferOwner(line: Line, ownerId: UUID?): WriteStatus {
        if (ownerId == null) {
            return WriteStatus.INVALID_VALUE
        }
        return if (lineManager.setLineOwner(line.id, ownerId)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun clearRoutePoints(line: Line): ClearRouteResult {
        val previousCount = line.routePoints.size
        val status = if (lineManager.clearLineRoutePoints(line.id)) WriteStatus.SUCCESS else WriteStatus.FAILED
        return ClearRouteResult(status, previousCount)
    }

    fun cloneReverseLine(sourceId: String, newId: String?, stopIdSuffix: String?, ownerId: UUID?): WriteStatus {
        if (!isValidId(newId)) {
            return WriteStatus.INVALID_ID
        }
        return if (lineManager.cloneReverseLine(sourceId, newId, stopIdSuffix, ownerId)) {
            WriteStatus.SUCCESS
        } else {
            WriteStatus.FAILED
        }
    }

    fun setTicketPrice(id: String, price: Double): WriteStatus {
        if (!java.lang.Double.isFinite(price) || price < 0.0) {
            return WriteStatus.INVALID_VALUE
        }
        return if (lineManager.setLineTicketPrice(id, price)) WriteStatus.SUCCESS else WriteStatus.FAILED
    }

    fun setPriceRule(id: String, mode: String, basePrice: Double, perUnit: Double?, maxPrice: Double?): WriteStatus {
        if (basePrice < 0.0 || (perUnit != null && perUnit < 0.0) || (maxPrice != null && maxPrice < 0.0)) {
            return WriteStatus.INVALID_VALUE
        }

        val pricingMode = try {
            PriceRule.PricingMode.valueOf(mode.uppercase(Locale.getDefault()))
        } catch (exception: IllegalArgumentException) {
            return WriteStatus.INVALID_VALUE
        }

        val line = lineManager.getLine(id) ?: return WriteStatus.NOT_FOUND
        val rule = PriceRule(pricingMode, basePrice)
        if (pricingMode == PriceRule.PricingMode.DISTANCE && perUnit != null) {
            rule.setPerBlockRate(perUnit)
        }
        if (pricingMode == PriceRule.PricingMode.INTERVAL && perUnit != null) {
            rule.setPerIntervalRate(perUnit)
        }
        if (maxPrice != null && maxPrice > 0.0) {
            rule.setMaxPrice(maxPrice)
        }

        line.priceRule = rule
        lineManager.saveConfig()
        return WriteStatus.SUCCESS
    }

    fun resetPriceRule(id: String): Boolean {
        val line = lineManager.getLine(id) ?: return false
        line.priceRule = null
        lineManager.saveConfig()
        return true
    }

    fun setLineStatus(id: String, status: String): WriteStatus {
        val lineStatus = LineStatus.fromConfig(status)
        if (lineStatus == LineStatus.NORMAL && !status.equals("normal", ignoreCase = true)) {
            return WriteStatus.INVALID_VALUE
        }

        val line = lineManager.getLine(id) ?: return WriteStatus.NOT_FOUND
        val oldStatus = line.getLineStatus()
        if (oldStatus == lineStatus) {
            return WriteStatus.SUCCESS
        }

        line.setLineStatus(lineStatus)
        Bukkit.getPluginManager().callEvent(LineStatusChangeEvent(line, oldStatus, lineStatus))
        lineManager.saveConfig()
        return WriteStatus.SUCCESS
    }

    fun addAlternativeRoute(id: String, altLineId: String): Boolean {
        val line = lineManager.getLine(id) ?: return false
        val result = line.addAlternativeRoute(altLineId)
        if (result) {
            lineManager.saveConfig()
        }
        return result
    }

    fun removeAlternativeRoute(id: String, altLineId: String): Boolean {
        val line = lineManager.getLine(id) ?: return false
        val result = line.removeAlternativeRoute(altLineId)
        if (result) {
            lineManager.saveConfig()
        }
        return result
    }

    fun setSuspensionMessage(id: String, message: String?): Boolean {
        val line = lineManager.getLine(id) ?: return false
        line.suspensionMessage = message
        lineManager.saveConfig()
        return true
    }

    fun isValidId(id: String?): Boolean =
        id != null &&
            id.isNotBlank() &&
            id.length <= MAX_ID_LENGTH &&
            ID_PATTERN.matcher(id).matches()

    fun isValidColor(color: String?): Boolean =
        color != null &&
            (
                LEGACY_COLOR_PATTERN.matcher(color).matches() ||
                    HEX_COLOR_PATTERN.matcher(color).matches()
                )

    companion object {
        private const val MAX_ID_LENGTH = 64
        private val ID_PATTERN: Pattern = Pattern.compile("[A-Za-z0-9_-]+")
        private val LEGACY_COLOR_PATTERN: Pattern = Pattern.compile("(?i)&[0-9A-FK-OR]")
        private val HEX_COLOR_PATTERN: Pattern = Pattern.compile("(?i)&#[0-9A-F]{6}")
    }
}
